plugins/parquet-store/parquet_store.cpp

```diff
@@ -0,0 +1,564 @@
+  auto handle_query = detail::overload{
+    [&](const query::count& count) {
+      if (count.mode == query::count::estimate)
+        die("logic error detected - count estimate should not load "
+            "partition");
+      std::shared_ptr<arrow::RecordBatch> batch{};
+      for (const auto& rb : record_batch_reader) {
+        if (rb.ok()) {
+          const auto slice = arrow_table_slice_builder::create(*rb);
+          auto result = count_matching(slice, expr, ids);
+          num_hits += result;
+          self->send(count.sink, result);
+        } else {
+          VAST_WARN("skipping record batch that failed to read: '{}'",
+                    rb.status());
+        }
+      }
+    },
+    [&](const query::extract& extract) {
+      if (extract.policy == query::extract::preserve_ids) {
+        return self->send(extract.sink, table_slice{});
+      }
+      for (const auto& rb : record_batch_reader) {
+        if (rb.ok()) {
+          const auto slice = arrow_table_slice_builder::create(*rb);
+          auto final_slice = filter(slice, expr, ids);
+          if (final_slice) {
+            num_hits += final_slice->rows();
+            self->send(extract.sink, *final_slice);
+          } else {
+            // TODO: error handling?
+            fmt::print(stderr, "tp;no slice today\n");
+          }
+        }
+      }
+    },
+  };
+  caf::visit(handle_query, query.cmd);
+  return num_hits;
+}
+
+std::shared_ptr<arrow::Schema> parse_arrow_schema_from_metadata(
+  const std::shared_ptr<parquet::FileMetaData>& parquet_metadata) {
+  if (!parquet_metadata)
+    return {};
+  auto arrow_metadata
+    = parquet_metadata->key_value_metadata()->Get("ARROW:schema");
+  if (!arrow_metadata.ok())
+    return {};
+  auto decoded = detail::base64::decode(*arrow_metadata);
+  auto schema_buf = std::make_shared<arrow::Buffer>(decoded);
+  arrow::ipc::DictionaryMemo dict_memo;
+  arrow::io::BufferReader input(schema_buf);
+  auto arrow_schema = arrow::ipc::ReadSchema(&input, &dict_memo);
+  if (!arrow_schema.ok())
+    return {};
+  return *arrow_schema;
+}
+
+caf::expected<std::shared_ptr<arrow::Table>>
+read_parquet_buffer(const chunk_ptr& chunk) {
+  const auto* ptr
+    = static_cast<const uint8_t*>(static_cast<const void*>(chunk->data()));
+  auto bufr = std::make_shared<arrow::io::BufferReader>(
+    ptr, detail::narrow_cast<int64_t>(chunk->size()));
+  auto parquet_reader = parquet::ParquetFileReader::Open(bufr);
+  auto arrow_schema
+    = parse_arrow_schema_from_metadata(parquet_reader->metadata());
+  std::unique_ptr<parquet::arrow::FileReader> file_reader{};
+  if (!parquet::arrow::OpenFile(bufr, arrow::default_memory_pool(),
+                                &file_reader)
+         .ok())
+    return caf::exit_reason::unhandled_exception;
+  std::shared_ptr<arrow::Table> table{};
+  if (!file_reader->ReadTable(&table).ok())
+    return caf::exit_reason::unhandled_exception;
+  return map_enum_dictionary(arrow_schema, table);
```

Comment:
This function definitely needs a better name, but I think the general idea works.

It'd be nice to see how much overhead this has. You can easily instrument by using a steady clock, e.g.,

```c++
const auto start = std::chrono::steady_clock::now();
// do stuff
const auto delta = std::chrono::steady_clock::now() - start;
VAST_DEBUG("someone did stuff in {}", data{std::chrono::duration_cast<duration>(delta)});
```

Reply:
renamed function to `align_table_to_schema`, which hopefully better captures the intend here.

I added the time tracking directly to the function.